Enable keycloak authentication to fire weather page WPS #193

Nkuruba opened the pull request:
WPS #193

DerekRoberts committed:
Initial commit

DerekRoberts committed:
Load default issue templates for bug and feature

DerekRoberts committed:
Jenkins deployment (#4)

* Initial commit for bcdk:jenkins

* config update

* Pipeline adjustment

DerekRoberts committed:
Issue Templates + Task request

DerekRoberts committed:
Update issue templates

DerekRoberts committed:
Update issue templates

DerekRoberts committed:
Update issue templates

DerekRoberts committed:
Update issue templates

DerekRoberts committed:
Apache License (#47)

DerekRoberts committed:
README template (#48)

Sybrand committed:
Added a glossary section to the README.md

Sybrand committed:
Removed duplicate glossary entry

Sybrand committed:
Added BUI/DC to glossary

Sybrand committed:
Moved to Wiki

Sybrand committed:
Modified glossary entry

Sybrand committed:
Removed []

jakemorr committed:
Update issue templates

jakemorr committed:
Update issue templates

jakemorr committed:
Update issue templates

added Epic Issue type

DerekRoberts committed:
Contributing.md (#71)

* Code of Conduct from https://www.contributor-covenant.org/
* Code of Conduct contact upgraded to Jake

DerekRoberts committed:
CONDUCT.md, not CONTRIBUTING.md

DerekRoberts committed:
Pipeline (#72)

* Initial pipeline commit

* Post @bcgov/generator-bcdk:jenkins generator

DerekRoberts committed:
Base MDs (#74)

* Cut down README.md filler

* Cleaned up README.md and added Contributing

DerekRoberts committed:
First job (#73)

repo-mountie[bot] committed:
Add missing compliance audit file (#1)

* Add compliance audit file

* Updating compliance audit file

DerekRoberts committed:
Very basic build (#81)

* Very basic build

* Copy and paste quick run

* Excluded .vscode in .gitignore

* Cleaned up temp build config

* Cleaned up temp deployment config

* Jenkins no longer hard coding commands

* Cut down bc and dc

* Renamed and moved demo templates

* Removed VERSION from bc

* Removed var

* Build now using script

* Script now runs from any path

* Simplified build script

* Deploy script

* Cleaned up script output

* Jenkinsfile deploying with script

* oc_build.sh cleaned up, defaults to dry run

* oc_build.sh - verify oc login

* Improved oc_build.sh output

* Improved oc_build.sh output further

* oc_build.sh overhauled to use commented help in script

* oc_deploy.sh given same treatment as oc_build.sh

* Jenkinsfile now full of stubs

* Auto-formatter

DerekRoberts committed:
Move and remove stale files (#136)

DerekRoberts committed:
Shared Scripts and Subtree (#137)

* Copied scripts from wps-web

* Shuffle files, folders and sourcing for subtree

* envars pulling NAME from git, have overrides, using BC/DC paths relative to subtree

* Script simplification and variable revision

Commit:
Updated git hub event script with a function to call clean up job when a PR is closed (#139)

Kyubinhan committed:
Use oc start-build after applying the template (#159)

* Use oc start-build after applying the template and remove for loop for following the building process

* Cancel non complete build and start new build using oc cancel-build & start-build

DerekRoberts committed:
Add ConfigMaps to Cleanup and Remove Stale Files (#157)

* Include configs maps in cleanup
* No longer need .jenkins/
* Pointed formatter at CONDUCT.md

DerekRoberts committed:
Subtree update (#162)

* Simple subtree update script

DerekRoberts committed:
oc_cleanup.sh typo (#163)

DerekRoberts committed:
Script changes moving from bash to sh for Coco OC Runner (#180)

Kyubinhan committed:
Script changes for explicit build and deploy without using any internal triggers (#200)

* Remove cancel-build command in oc_build.sh

* Use oc rollout to deploy apps in oc_deploy.sh

Commit:
Added a new fire weather page and keycloak libraries for enabling IDIR based authentication

Commit:
Removed unnecessary imports

Commit:
Using hooks to manage state and rendering of components

Commit:
Merging from master

Commit:
Added config to enable authentication for accessing fire weather page

Commit:
Removed unnecessary space

Commit:
Updated readme with extra step

Commit:
Updated gitignore

Commit:
Updated readme and added keycloak example file

Commit:
Added react testing lib to prod dependencies to resolve the issue

Commit:
Enhancement to menu items

Sybrand commented on src/components/PageHeader.tsx:
```diff
@@ -59,14 +60,24 @@ export const PageHeader = ({ title }: Props) => {
-        </a>
+        <div>
+          <a
+            className={classes.contact}
```
We haven't decided that we're going to have a link to the fire weather calculator, and it's not part of the acceptance criteria of this story.

Nkuruba replied:
@Sybrand the link is for temporary purpose as it would be removed as part of next coming PRs. The link was only introduced to showcase the navigation and login functionality.

Sybrand replied:
We can demonstrate this by linking directly to a url know to us - please remove it from this PR as it ads unwanted functionality.

Nkuruba replied:
ok will remove it.

Sybrand commented on src/utils/keycloak.example.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "initOptions": {
+    "url": "https://sso.pathfinder.gov.bc.ca/auth",
```
This has to be different for dev/test/prod - so it can't be in a hard coded json file. Please move to environment variables.

Nkuruba replied:
@Sybrand this is an example file and it was added only to run the app locally.

Kyubinhan replied:
Can you then change it to `https://sso-dev.pathfinder.gov.bc.ca/auth`

Nkuruba replied:
ok

Sybrand replied:
My bad - sorry @Nkuruba

Sybrand commented on package.json:
```diff
@@ -17,15 +17,19 @@
+    "@types/react-router-dom": "^5.1.3",
     "@types/webpack-env": "^1.15.1",
     "axios": "^0.19.2",
+    "keycloak-js": "^9.0.2",
```
Does this pull keycloak.js from https://sso.pathfinder.gov.bc.ca/auth/js/keycloak.js ? This is important, as the version of the client has to match that of the server.

Nkuruba replied:
@Sybrand no but react has an official keycloak library to manage calls to keycloak server.

Sybrand replied:
If the keycloak client does not match the keycloak server version - there will be issues. This was clearly specified in the acceptance criteria, and clearly stated as best practice on the official documentation.

For more detail, please refer to the official documentation: https://www.keycloak.org/docs/latest/securing_apps/#_javascript_adapter

"A best practice is to load the JavaScript adapter directly from Keycloak Server as it will automatically be updated when you upgrade the server. If you copy the adapter to your web application instead, make sure you upgrade the adapter only after you have upgraded the server."

Nkuruba replied:
@Sybrand added code to read from keycloak.js on the fly.

Kyubinhan commented on src/components/PageHeader.tsx:
```diff
@@ -59,14 +60,24 @@ export const PageHeader = ({ title }: Props) => {
-          Contact
-        </a>
+        <div>
+          <a
+            className={classes.contact}
+            href="/fireWeatherCalculator"
+            target="_blank"
+            rel="noopener noreferrer"
+          >
+            FireWeather
+          </a>
```
Do you think we need this @Sybrand ? Cause my understanding was each product(pages) are gonna be independent

Sybrand replied:
We explicitly agreed that they were going to be independent, so agreed - I don't know where this came from.

Nkuruba replied:
@Sybrand @Kyubinhan yes fire weather is a independent component. I just added this link to as as a temporary solution. This will be removed when actual routing would be added to the application. There is already another ticket for this purpose wps #196.

Nkuruba replied:
@Sybrand @Kyubinhan I removed the link.

Nkuruba commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -40,12 +40,26 @@ parameters:
+  - name: KEYCLOAK_REALM
+    value: 8wl6x4cp
+  - name: KEYCLOAK_CLIENT
+    value: security-admin-console
```
@Kyubinhan yes but it was not part of acceptance criteria so I managed to use the current one. We can create new one as part of a different ticket.

Sybrand replied:
@Kyubinhan do you mean a new realm entirely?

Kyubinhan replied:
@Sybrand No. Clients are part sso realms.

Sybrand replied:
@Kyubinhan phew - ok - yes - agreed, we need a different client. And no, @Nkuruba not as part of a separate PR. This is not the security-admin-console, this is something else.

Nkuruba replied:
@Sybrand @Kyubinhan created wps-web client in all envs.

Kyubinhan commented on src/features/fwCalculator/FWCalculatorPage.tsx:
```diff
@@ -0,0 +1,25 @@
+import { initOptions } from 'utils/keycloak.json'
+
+export const FWCalculatorPage = () => {
+  const [isAuthenciated, setAuthenciation] = React.useState(false)
+  React.useEffect(() => {
+    const keycloak = Keycloak(initOptions)
+    keycloak
+      .init({ onLoad: 'login-required', checkLoginIframe: false })
+      .then(authenticated => {
+        if (!authenticated) {
+          window.location.reload()
+        } else {
+          setAuthenciation(authenticated)
+        }
+        setTimeout(() => {
+          keycloak.updateToken(60)
+        }, 60000)
+      })
+  }, [])
+
+  if (isAuthenciated) return <header>Hello World!!!</header>
+  else return <div />
```
We need to store `isAuthenciated` in Redux not in local state

Nkuruba replied:
@Kyubinhan the authentication state is not being stored in redux store

Nkuruba commented on package.json:
```diff
@@ -17,15 +17,19 @@
     "@material-ui/icons": "^4.9.1",
     "@material-ui/lab": "^4.0.0-alpha.43",
     "@reduxjs/toolkit": "^1.2.5",
+    "@testing-library/react": "^9.5.0",
```
@Kyubinhan the app was not starting as it was missing this library and I had to add it to resolve the issue.

Sybrand replied:
Please resolve this issue without including developer dependancies.

Nkuruba commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -40,12 +40,26 @@ parameters:
     description: Memory upper limit
     value: 2Gi
+  - name: KEYCLOAK_URL_AUTH
+    value: https://sso.pathfinder.gov.bc.ca/auth
```
@Kyubinhan this is a pipeline design related issue as we are not following different URL's for API as well. We are using prod API URL by default in the template and I had to follow same approach. We can discuss on this and implement it as part of a new ticket to separate out config for different environment.

Sybrand replied:
Agreed - same as per json file.

Sybrand replied:
i.m.h.o - this has to be solved as part of this PR.

Nkuruba replied:
@Sybrand @Kyubinhan code added to handle this.

Nkuruba commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -40,12 +40,26 @@ parameters:
   - name: MEMORY_LIMIT
     description: Memory upper limit
     value: 2Gi
+  - name: KEYCLOAK_URL_AUTH
```
@Kyubinhan my idea of using this variable was to convey that it is keycloak url for authentication so I want to stick with  KEYCLOAK_URL_AUTH.

Sybrand replied:
Agreed.

Nkuruba replied:
ok

Nkuruba commented on src/app/App.tsx:
```diff
@@ -40,7 +42,18 @@ const App = () => {
+              <FWICalculatorPage />
+            )}
+          </Route>
+          <Route path="/fireWeatherCalculator" exact>
```
ok

Nkuruba commented on src/app/App.tsx:
```diff
@@ -40,7 +42,18 @@ const App = () => {
       <ThemeProvider theme={theme}>
-        {showModal ? <FWICalculatorDisclaimerModal /> : <FWICalculatorPage />}
+        <Router>
+          <Route path="/" exact>
```
@Kyubinhan why do you need to redirect users from calculator page?

Commit:
Updated fire weather page to use keycloak js file on the fly

Commit:
Updated branch with changes from master

Commit:
Added authentication state to redux store rather than to local state

Commit:
Updated keykloak client name

Commit:
Added a test for testing authentication slice

Commit:
Updated template parameter name

Commit:
Updated package lock

Commit:
Updated package lock

Commit:
Updated package and package-lock

Commit:
Added reset function to reset the authentication state

Commit:
Added an env var to update keycloak auth url when promoting to prod

Commit:
Added keycloak dev URL to env vars

DerekRoberts commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -40,12 +40,27 @@ parameters:
     value: 2Gi
+  - name: KEYCLOAK_AUTH_URL
+    description: Keycloak url for authentication
+    required: true
```
Please replace `required: true` with `value: https://DEV_ADDRESS`.

Commit:
Added dev url as a default value

Kyubinhan commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -72,6 +87,10 @@ objects:
               imagePullPolicy: Always
+              volumeMounts:
+                - name: ${NAME}-${SUFFIX}
+                  mountPath: /app/src/utils/keycloak.json
```
Can you move 3 variables stored in the `keycloak.json` into `.env` file and use them just like how we use other env variables. I don't see any reason that they are being treated differently.

Nkuruba replied:
@Kyubinhan I dont want to introduce more environment variables. Also please go through https://www.keycloak.org/docs/latest/securing_apps/#_javascript_adapter as I am using same concept.

Kyubinhan replied:
I don't see anywhere not to use environment variables in this docs. You can still initiate keycloak instance the way you want, but I am suggesting to reference env variables because I don't want to add a file coming from a configmap in build time later when we go production mode. You can mount it like this `/app/src/utils/keycloak.json` because we are running the app in dev mode

Nkuruba replied:
@Kyubinhan the config map will be mounted during deployment time rather build time. Also I have already suggested that there may be more params coming in as part of Keycloak config and I dont want whole lot of env variables floating everywhere.

Kyubinhan replied:
Currently yes the config map is mounted during deployment, but your strategy won't work if we build the app into static files during the build time when we go to prod using caddy. Take a good look at Devhub's DC and BC because this is what we are aiming for. https://github.com/bcgov/devhub-app-web/blob/master/openshift/templates/dc.yaml

Kyubinhan commented on openshift/scripts/common/envars:
```diff
@@ -8,3 +8,5 @@ PROJ_PROD="${PROJ_PROD:-auzhsi-prod}"
 TAG_PROD="${TAG_PROD:-prod}"
 PATH_BC="${PATH_BC:-$(dirname ${0})/../templates/build.bc.yaml}"
 PATH_DC="${PATH_DC:-$(dirname ${0})/../templates/deploy.dc.yaml}"
+KEYCLOAK_PROD_URL="https://sso.pathfinder.gov.bc.ca/auth"
+KEYCLOAK_DEV_URL="https://sso-dev.pathfinder.gov.bc.ca/auth"
```
I would name them SSO_BASE_PROD_URL and SSO_BASE_DEV_URL and btw I think you can just set the dev url here and use `https://sso.pathfinder.gov.bc.ca/auth` as a default in the template

Nkuruba replied:
@Kyubinhan I want to stick with KEYCLOAK naming convention rather than SSO. Also I dont want app to point to prod by default.

DerekRoberts reviewed (dismissed):
Anything that touches deployment and the pipeline looks good!  :)

Commit:
Updated test description

Kyubinhan commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,46 @@
+  }, [dispatch])
+
+  if (isAuthenticated) return <header>Hello World!!!</header>
+  else return <div />
```
Can you render a message showing what state users are in? Rendering an empty <div> tag is not user-friendly.

Nkuruba replied:
@Kyubinhan I am not routing users to empty page with just <Div/>. This is just on way login flow.

Kyubinhan replied:
I am not talking about routing. I am pointing out that in this if condition there is a change to render an empty div tag.

Kyubinhan commented on src/features/fireWeather/slices/authenticationSlice.ts:
```diff
@@ -0,0 +1,58 @@
+
+export default auth.reducer
+
+export const createAuthentication = (
+  authenticated: boolean
+): AppThunk => async dispatch => {
+  try {
+    dispatch(getAuthenticationStart())
+    dispatch(getAuthenticationSuccess(authenticated))
+  } catch (err) {
+    dispatch(getAuthenticationFailed(err.toString()))
+  }
+}
```
I don't see any purpose of this action creator because there is no side effect involved. Side effect is actually happening when you run promise in `FireWeatherPage.jsx`

Nkuruba replied:
@Kyubinhan I placed it thinking we can use it if there were any side effects later so that no need to write this function entirely from scratch. Anyway I have removed it.

Kyubinhan commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,46 @@
+            dispatch(resetAuthentication())
+            window.location.reload()
+          } else {
+            dispatch(createAuthentication(authenticated))
```
Dispatch getAuthenticationSuccess here. More explanation written in the slice file

Nkuruba replied:
@Kyubinhan Updated it.

Kyubinhan commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,46 @@
+          setTimeout(() => {
+            keycloak.updateToken(60)
+          }, 6000)
+        })
```
Use `.catch` here if you want to dispatch `getAuthenticationFailed` More explanation written in the slice file

Nkuruba replied:
@Kyubinhan added catch.

Kyubinhan commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,46 @@
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    const script = document.createElement('script')
+
+    script.src = `${initOptions.url}/js/keycloak.js`
+    script.async = true
+
+    const keyCloakScript = document.body.appendChild(script)
```
You are applying this script every time useEffect is being run, so move this out from React related stuff

Nkuruba replied:
@Kyubinhan done

Commit:
1. Removed app thunk 2. Using env variables instead of config map 3. Loading keycloak.js independently from react component

Commit:
Removed unused import

Commit:
Removed unused volume mount

Kyubinhan commented on public/index.html:
```diff
@@ -20,6 +20,8 @@
     <link rel="icon" href="%PUBLIC_URL%/images/bcid-favicon-16x16.png" sizes="16x16" type="image/png" />    
     <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />
 
+    <script src="%REACT_APP_KEYCLOAK_AUTH_URL%/js/keycloak.js"></script>
```
Nice!

Kyubinhan commented on .gitignore:
```diff
@@ -78,4 +78,7 @@ jspm_packages/
\ No newline at end of file
+.vscode/*
+
+# keycloak json
+keycloak.json
```
Can you remove these lines?

Kyubinhan commented on README.md:
```diff
@@ -22,7 +22,7 @@ Installs all dependencies in the node_modules folder.
 ### Executing program
 
-In the project directory, create `.env` file at root using `.env.example` as a sample, then you can run:
+In the project directory, create `.env` file at root using `.env.example` as a sample and then you can run:
```
Can you revert this change?

Kyubinhan commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,50 @@
+            dispatch(getAuthenticationFailed(err))
+          }
+        }, 6000)
+      })
```
What I was asking for was actually using Promise chaining.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises.
So right after then() block you can add .catch

Kyubinhan reviewed (changes requested):
Few more changes please

DerekRoberts committed:
Promote to distinct PROD imagestream, retage explicitly

Commit:
Updated gitignore and readme

DerekRoberts committed:
Removed common tag from now separate imagestreams

Sybrand commented on openshift/scripts/common/envars:
```diff
@@ -8,3 +8,5 @@ PROJ_PROD="${PROJ_PROD:-auzhsi-prod}"
 TAG_PROD="${TAG_PROD:-prod}"
 PATH_BC="${PATH_BC:-$(dirname ${0})/../templates/build.bc.yaml}"
 PATH_DC="${PATH_DC:-$(dirname ${0})/../templates/deploy.dc.yaml}"
+KEYCLOAK_PROD_URL="https://sso.pathfinder.gov.bc.ca/auth"
```
@Nkuruba @Kyubinhan we're being silly (re. our discussion about this yesterday) - and sorry if I missed the plot yesterday - but we don't need this at all. The DC should be pulling this from a configmap - so we don't have to specify the url here at all.

Sybrand commented on openshift/scripts/oc_deploy.sh:
```diff
@@ -21,13 +21,20 @@ source "$(dirname ${0})/common/common"
 #%   ${THIS_FILE} pr-0 apply
 #%
 
+# Use keycloak prod url when deploying to prod
```
@Nkuruba @Kyubinhan again - we don't need this here at all - we don't need to pass it to the template, the template can grab the values from the configmap.

Nkuruba replied:
@Sybrand now the template is grabbing values from global configmap.

Sybrand commented on openshift/scripts/oc_deploy.sh:
```diff
@@ -21,13 +21,20 @@ source "$(dirname ${0})/common/common"
 # Process a template (mostly variable substition)
 #
-OC_PROCESS="oc -n ${PROJ_TOOLS} process -f ${PATH_DC} -p NAME=${NAME_APP} -p SUFFIX=${SUFFIX}"
+OC_PROCESS="oc -n ${PROJ_TOOLS} process -f ${PATH_DC} -p NAME=${NAME_APP} -p SUFFIX=${SUFFIX} -p KEYCLOAK_AUTH_URL=${KEYCLOAK_AUTH_URL}"
```
Please remove this - sorry for asking for it yesterday - I was totally wrong - this doesn't have to be passed to the template - we need a global configmap in our dev project that has all of these things specified.

Sybrand commented on src/features/fireWeather/FireWeatherPage.jsx:
```diff
@@ -0,0 +1,49 @@
+  const dispatch = useDispatch()
+
+  useEffect(() => {
+    const keycloak = window.Keycloak({
```
This is a good solution to start with - we'll need to revisit the moment we have more than one page needing logins, and I suspect we're not dealing correctly with tokens going stale etc. but that's a problem for another day.

Sybrand commented on public/index.html:
```diff
@@ -20,6 +20,8 @@
     <link rel="icon" href="%PUBLIC_URL%/images/bcid-favicon-16x16.png" sizes="16x16" type="image/png" />    
     <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />
 
+    <script src="%REACT_APP_KEYCLOAK_AUTH_URL%/js/keycloak.js"></script>
```
Fine for now, but we're going to have to solve async loading at some point - I'll write a ticket for it.

Kyubinhan replied:
By putting this script tag in `<head>`tag we don't have to worry about loading issue because this will be loaded first before React related scripts

Sybrand replied:
It is going to slow down the overall loading of the page though. On second thought, it's not even worth a ticket - but we can get faster loading overall by loading it asynchronously and putting the wait in where we're actually needing to auth. BUT - REALLY - the file is going to be cached by the browser, so never mind - too much effort for speeding up the initial load.

Sybrand reviewed (changes requested):
Some comments - but only one change we need:
We can rely on config maps - we don't need to pollute things with dev/prod environment variables - sorry that I missed that yesterday when we walked through the code.

Commit:
Using global config map to fetch env vars

DerekRoberts committed:
Remove scripts dir, to replace

DerekRoberts committed:
Merge commit '1780e0ead79a9c4ec866bf37455e052ca0f47dba' as 'openshift/scripts'

Commit:
Updated config map name

Nkuruba commented:
> Some comments - but only one change we need:
> We can rely on config maps - we don't need to pollute things with dev/prod environment variables - sorry that I missed that yesterday when we walked through the code.

@Sybrand now the deployment template uses values from global config map.

Sybrand reviewed (changes requested):
Something seems to have gone wrong with deployment, I suspect you're going to have to close this PR and do a new one. (I think the failed production push did some cleanup that's now causing the build to fail?)

Nkuruba commented:
> Something seems to have gone wrong with deployment, I suspect you're going to have to close this PR and do a new one. (I think the failed production push did some cleanup that's now causing the build to fail?)

@Sybrand could you please elaborate? where is the build failing?

Kyubinhan commented on openshift/templates/deploy.dc.yaml:
```diff
@@ -40,16 +36,9 @@ parameters:
   - name: MEMORY_LIMIT
     description: Memory upper limit
     value: 2Gi
+  - name: WPS_WEB_GLOBAL_PROPS
```
Can you rename this to `WPS_WEB_GLOBAL_CONFIG` or `WPS_WEB_GLOBAL_CONFIGMAP`

Nkuruba replied:
done

Commit:
Renamed variable name

Commit:
Merge remote-tracking branch 'origin/master' into WPS-193

DerekRoberts committed:
Bug - use SUFFIX, not latest for tags

DerekRoberts committed:
Bug - use SUFFIX, not latest for tags

DerekRoberts committed:
Replacing script subtree

DerekRoberts committed:
Add 'openshift/scripts/' from commit '5ec077027f782aef715d5f83a48c5b22ab89b7b3'

git-subtree-dir: openshift/scripts
git-subtree-mainline: 325581e7711333a97985ac5676a38e24a9fe8e09
git-subtree-split: 5ec077027f782aef715d5f83a48c5b22ab89b7b3

DerekRoberts reviewed (approved):
I think this is ready to go!  Good job surviving us.  ;)
